instrumentation/netty/netty-4-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/netty/common/client/AbstractNettyHttpClientTracer.java

```diff
@@ -84,39 +89,58 @@ protected String responseHeader(HttpResponse httpResponse, String name) {
     return SETTER;
   }
 
-  public static void operationComplete(Future<?> future) {
-    AbstractNettyHttpClientTracer tracer = NettyHttpClientTracerAccess.getTracer();
-    if (tracer == null) {
-      return;
+  public Context startConnectionSpan(Context parentContext, SocketAddress remoteAddress) {
+    if (!alwaysCreateConnectSpan) {
+      return null;
     }
 
-    if (!(future instanceof ChannelFuture)) {
-      return;
-    }
-    // If first call to GenericFutureListener#operationComplete has an exception then we
-    // treat it as the cause of connection failure and create a special span for it
-    ChannelFuture channelFuture = (ChannelFuture) future;
-    Context parentContext = tracer.getAndRemoveConnectContext(channelFuture);
-    if (parentContext == null) {
-      return;
-    }
-    Throwable cause = future.cause();
-    if (cause == null) {
-      return;
-    }
+    SpanBuilder spanBuilder = spanBuilder(parentContext, "CONNECT", INTERNAL);
+    NetPeerAttributes.INSTANCE.setNetPeer(spanBuilder, (InetSocketAddress) remoteAddress);
+
+    return parentContext.with(spanBuilder.startSpan());
+  }
 
-    if (tracer.shouldStartSpan(parentContext, SpanKind.CLIENT)) {
-      tracer.connectionFailure(parentContext, channelFuture.channel(), cause);
+  public void endConnectionSpan(
+      Context context,
+      Context parentContext,
+      SocketAddress remoteAddress,
+      Channel channel,
+      Throwable throwable) {
+    if (context != null) {
+      // if context is present we started span in startConnectionSpan
+      endConnectionSpan(context, channel, throwable);
+    } else if (throwable != null && shouldStartSpan(parentContext, CLIENT)) {
```

Comment:
As `connectionFailure` creates INTERNAL span, I don't think we need `shouldStartSpan` here?

Reply:
In case of a http client based on netty that has custom instrumentation that captures both connection creation and actual request there is no need need to create the connection span just for failure, the same information will be recorded under parent span.